### PR TITLE
Query bank balances

### PR DESCRIPTION
**Pull request by samsiegart**
refs https://github.com/Agoric/wallet-app/issues/49

We still need to update dapp-inter to do the same, then I can test e2e with amounts in offer proposals rather than pursepetname+value

### CODE REVIEW

File: src/util/WalletBackendAdapter.ts [hunk 1]
```diff
@@ -188,15 +205,79 @@ export const makeWalletBridgeFromFollowers = (
...
+
+    const possiblyUpdateBankPurses = () => {
+      if (!vbankAssets || !bank) return;
+
+      const bankMap = new Map<string, string>(
+        bank.map(({ denom, amount }) => [denom, amount]),
+      );
+
+      vbankAssets.forEach(([denom, info]) => {
+        const amount = bankMap.get(denom) ?? 0n;
```
Comment (samsiegart) [inline src/util/WalletBackendAdapter.ts:233]: If an asset exists in the vbank we want to show it as a purse with 0 balance so they can click the eventual "deposit" button for ibc transfer.

File: src/service/Offers.ts [hunk 2]
```diff
@@ -41,33 +39,43 @@ export const getOfferService = (
...
         Object.entries(paymentProposals).map(
           // @ts-expect-error
-          ([kw, { pursePetname, value }]) => {
-            const brand = pursePetnameToBrand.get(pursePetname);
-            if (!brand) {
+          async ([kw, { pursePetname, value, amount: serializedAmount }]) => {
+            /// XXX test e2e with dapp inter once feasible.
+            const amount = serializedAmount
+              ? await E(boardIdMarshaller).unserialize(serializedAmount)
+              : { brand: pursePetnameToBrand.get(pursePetname), value };
```
Comment (dckc) [inline src/service/Offers.ts:53]: we're cool with open-coding `AmountMath.make()`? maybe the notion of `brand` used here doesn't quite line up?

File: src/util/queryBankBalances.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import { Tendermint34Client } from '@cosmjs/tendermint-rpc';
+import { QueryClientImpl } from 'cosmjs-types/cosmos/bank/v1beta1/query';
+import type { Coin } from '@cosmjs/stargate';
+import type { HttpEndpoint } from '@cosmjs/tendermint-rpc';
+
+export const queryBankBalances = async (
+  address: string,
+  rpc: HttpEndpoint,
+): Promise<Coin[]> => {
+  const tendermint = await Tendermint34Client.connect(rpc);
```
Comment (dckc) [inline src/util/queryBankBalances.ts:11]: ambient authority. hm.

> No modules should export powerful objects (for example, objects that close over fs or process.env). -- https://github.com/Agoric/agoric-sdk/wiki/OCap-Discipline

I suppose we have agreed on an exception to that for UI code? I'd like to see that in the README or the like.

File: src/util/queryBankBalances.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  const queryClient = new QueryClient(tendermint);
+  const rpcClient = createProtobufRpcClient(queryClient);
+  const bankQueryService = new QueryClientImpl(rpcClient);
+
+  const { balances } = await bankQueryService.AllBalances({
```
Comment (dckc) [inline src/util/queryBankBalances.ts:16]: **scaling consideration**: How often do we do this? I gather from @arirubinstein that we're supposed to make no more than 2 RPC queries per second.

I gather @michaelfig plans to provide rate limiting in `@agoric/casting` (IOU an issue #) Do we have an issue about migrating this code to use it? Cite it from a TODO?

File: src/util/WalletBackendAdapter.ts [hunk 5]
```diff
@@ -210,13 +291,29 @@ export const makeWalletBridgeFromFollowers = (
...
-        bd => purse.brand === bd.brand,
-      );
-      assert(brandDescriptor, `missing descriptor for brand ${purse.brand}`);
+      // We only care about the zoe invite purse, which has asset kind 'set'.
+      // Non 'set' amounts need to be fetched from vbank to know their
+      // decimalPlaces.
+      //
+      // If we have add non 'set' amount purses that aren't in the vbank, it's
+      // not currently possible to read their decimalPlaces, so this code
+      // will need updating.
```
Comment (dckc) [inline src/util/WalletBackendAdapter.ts:306]: This documentation is good.

Ideally we'd import some `invitationIssuerAssetKind` constant from `@agoric/zoe/something`.

I leave it to @turadg to judge the cost-effectiveness of that.

**Review by dckc — COMMENTED** [review]
Looks plausible.

I'd like to see test coverage for this feature, as usual. Skipping it for now is OK with me if it's OK with @turadg .

File: src/util/queryBankBalances.ts (hunk 4 above)
Comment (samsiegart) [reply src/util/queryBankBalances.ts:16]: Yea my hope is this could unblock us for now and we can swap in the improved casting stuff later, just added a comment here.

File: src/util/queryBankBalances.ts (hunk 3 above)
Comment (samsiegart) [reply src/util/queryBankBalances.ts:11]: Noted, I don't know if I have a clear way to explain the philosophy as it pertains to UI code at the moment though.

File: src/service/Offers.ts (hunk 2 above)
Comment (samsiegart) [reply src/service/Offers.ts]: Thanks, you caught a bug, I wasn't converting value to BigInt. Just changed to use AmountMath and confirmed it works.

File: src/service/Offers.ts [hunk 6]
```diff
@@ -41,33 +40,47 @@ export const getOfferService = (
...
       instanceHandle,
       publicInvitationMaker,
-      proposalTemplate: { give, want },
+      proposalTemplate: { give: giveTemplate, want: wantTemplate },
     } = offer;
 
-    const mapPursePetnamesToBrands = paymentProposals =>
-      Object.fromEntries(
+    const convertProposals = async paymentProposals => {
+      const entries = await Promise.all(
```
Comment (turadg) [inline src/service/Offers.ts:47]: this is a little hard to follow. consider filtering and naming the value transformer.

```
const convertProposal = ({ pursePetname, value, amount: serializedAmount }) =>
  serializedAmount
              ? await E(boardIdMarshaller).unserialize(serializedAmount)
              : AmountMath.make(
                  pursePetnameToBrand.get(pursePetname),
                  BigInt(value);

Object.entries(paymentProposals).
filter(([_kw, { pursePetname, amount}]) => amount && pursePetnameToBrand.get(pursePetname))
.map(([kw, proposal]) => convertProposal);
```

Then it's a small step to using `objectMap` and `deeplyFulfilled`;

File: src/service/Offers.ts [hunk 7]
```diff
@@ -41,33 +40,47 @@ export const getOfferService = (
...
             }
-            return [
-              kw,
-              {
-                brand,
-                value: BigInt(value),
-              },
-            ];
+
+            /// XXX test e2e with dapp inter once feasible.
```
Comment (turadg) [inline src/service/Offers.ts:55]: thanks for the documenting. consider TODO.

File: src/service/Offers.ts [hunk 8]
```diff
@@ -41,33 +40,47 @@ export const getOfferService = (
...
 
-    const instance = await E(boardIdMarshaller).unserialize(instanceHandle);
-    const {
-      slots: [instanceBoardId],
-    } = await E(boardIdMarshaller).serialize(instance);
+    const deconstructInstance = async () => {
+      const instance = await E(boardIdMarshaller).unserialize(instanceHandle);
+      const {
+        slots: [instanceBoardId],
+      } = await E(boardIdMarshaller).serialize(instance);
```
Comment (turadg) [inline src/service/Offers.ts:74]: i had to remember there's no promise pipelining in this app so the two awaits are fine.

it wasn't obvious to me why you wanted to wrap these in a function. It looks like you want `convertProposals` to resolve in parallel to resolving the `instance`. Good idea. But I don't see why you want `instanceBoardId` to be available after this Promise.all.

consider,
```
    const [instance, give, want] = await Promise.all([
      E(boardIdMarshaller).unserialize(instanceHandle),
      convertProposals(giveTemplate),
      convertProposals(wantTemplate),
    ]);
```
and then just above the first use of `instanceBoardId`,
```
    const {
      slots: [instanceBoardId],
    } = await E(boardIdMarshaller).serialize(instance);
```

File: src/util/WalletBackendAdapter.ts [hunk 9]
```diff
@@ -188,15 +205,79 @@ export const makeWalletBridgeFromFollowers = (
...
+    makeNotifierKit<Number | null>(null);
+
   const watchBeansOwing = async () => {
     for await (const { value } of iterateLatest(beansOwingFollower)) {
+      if (isHalted) return;
       beansOwingUpdater.updateState(Number(value));
     }
   };
 
+  // Reads purses from the cosmos bank module. These are not necessarily real
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:218]: consider "Infer purses from Cosmos bank module balances" since, as you say, purses can't be read from Cosmos.

File: src/util/WalletBackendAdapter.ts (hunk 1 above)
Comment (turadg) [reply src/util/WalletBackendAdapter.ts:233]: Very helpful note. Please include in the code.

File: src/util/WalletBackendAdapter.ts [hunk 10]
```diff
@@ -188,15 +205,79 @@ export const makeWalletBridgeFromFollowers = (
...
+      });
+
+      isBankLoaded = true;
+      updatePurses();
+    };
+
+    const watchBank = async () => {
+      if (isHalted) return;
+      const balances = await queryBankBalances(keplrConnection.address, rpc);
+      bank = balances;
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:251]: why the const?
```suggestion
      bank = await queryBankBalances(keplrConnection.address, rpc);
```

File: src/util/WalletBackendAdapter.ts [hunk 11]
```diff
@@ -188,15 +205,79 @@ export const makeWalletBridgeFromFollowers = (
...
+      }
+    };
+
+    void watchVbankAssets();
+    void watchBank();
+  };
+
+  const fetchAgoricBrands = async () => {
+    for await (const { value } of iterateLatest(agoricBrandsFollower)) {
+      if (value) {
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:270]: when would the iterator return an object without a `value`?

File: src/util/WalletBackendAdapter.ts [hunk 12]
```diff
@@ -188,15 +205,79 @@ export const makeWalletBridgeFromFollowers = (
...
+        return new Map(value.map(([k, v]) => [v, k]));
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:271]: This took me some time to read through without type info. I think it's making a map with the keys/values inverted. Consider https://lodash.com/docs/4.17.15#invert to make that clear, or a code comment.

File: src/util/WalletBackendAdapter.ts [hunk 13]
```diff
@@ -120,13 +125,20 @@ export const makeWalletBridgeFromFollowers = (
   currentFollower: ValueFollower<CurrentWalletRecord>,
   updateFollower: ValueFollower<UpdateRecord>,
   beansOwingFollower: ValueFollower<string>,
+  vbankAssetsFollower: ValueFollower<unknown>,
+  agoricBrandsFollower: ValueFollower<unknown>,
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:129]: please include the types. if they can't be imported just write out what this code is assuming.

Comment (turadg) [reply src/util/WalletBackendAdapter.ts:129]: also this is getting very long for a positional arguments list. very much worth turning into a named options argument, though not necessarily this PR

File: src/util/WalletBackendAdapter.ts (hunk 5 above)
Comment (turadg) [reply src/util/WalletBackendAdapter.ts:306]: I don't know of any runtime features of an invitation issuer that distinguishes it from any other Set kind issuer.

You could make this "set"-ness check more apparent using `@agoric/patterns` (soon to be in Endo): `matches(purse.balances.value, SetValueShape)`.

File: src/util/WalletBackendAdapter.ts [hunk 14]
```diff
@@ -210,13 +291,29 @@ export const makeWalletBridgeFromFollowers = (
...
+      //
+      // If we have add non 'set' amount purses that aren't in the vbank, it's
+      // not currently possible to read their decimalPlaces, so this code
+      // will need updating.
+      if (
+        !Array.isArray(purse.balance.value) ||
+        !agoricBrands.has(purse.brand)
+      ) {
+        continue;
+      }
```
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:312]: consider distinguishing these conditions:
```
if (!Array.isArray(purse.balance.value)) {
  console.debug('skipping non-set amount', purse.balance.value);
}
if (!agoricBrands.has(purse.brand)) {
  console.debug('skipping unknown brand', purse.brand);
```
Likely to pay off debugging someday. And in the meantime it's code documentation.

File: src/util/queryBankBalances.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import { QueryClient, createProtobufRpcClient } from '@cosmjs/stargate';
+import { Tendermint34Client } from '@cosmjs/tendermint-rpc';
+import { QueryClientImpl } from 'cosmjs-types/cosmos/bank/v1beta1/query';
+import type { Coin } from '@cosmjs/stargate';
+import type { HttpEndpoint } from '@cosmjs/tendermint-rpc';
+
+// XXX use casting support when possible for load balancing, batching, and
```
Comment (turadg) [inline src/util/queryBankBalances.ts:7]: Do we have a ticket to track when it's possible? I think we should. Consider noting it here with an `UNTIL`.

File: src/util/queryBankBalances.ts (hunk 3 above)
Comment (turadg) [reply src/util/queryBankBalances.ts:13]: I think we have agreed agreed on an exception to that for UI code and I agree it would be valuable to have in the README of this repo to refer to. Not a blocker imo

File: yarn.lock [hunk 16]
```diff
@@ -12164,7 +12164,7 @@ rollup@^2.43.1:
   optionalDependencies:
     fsevents "~2.3.2"
 
-rollup@endojs/endo#rollup-2.7.1-patch-1, "rollup@github:endojs/endo#rollup-2.7.1-patch-1":
+rollup@endojs/endo#rollup-2.7.1-patch-1:
```
Comment (turadg) [inline yarn.lock:12167]: curious.

File: src/service/Offers.ts (hunk 8 above)
Comment (samsiegart) [reply src/service/Offers.ts]: Done.

I'm not sure about promise pipelining in the rest of the app, but we await here so that the card doesn't show up before it has the data needed to render.

File: src/service/Offers.ts (hunk 7 above)
Comment (samsiegart) [reply src/service/Offers.ts]: Done

File: src/service/Offers.ts (hunk 6 above)
Comment (samsiegart) [reply src/service/Offers.ts:47]: I gave this a shot but I got a little confused with how to use those utility functions and the code wasn't looking much simpler, I'd prefer to leave this as is.

File: src/util/WalletBackendAdapter.ts (hunk 13 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Added types here and a bit below.

File: src/util/WalletBackendAdapter.ts (hunk 9 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Done

File: src/util/WalletBackendAdapter.ts (hunk 1 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts:244]: Done

File: src/util/WalletBackendAdapter.ts (hunk 10 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Oops, done

File: src/util/WalletBackendAdapter.ts (hunk 11 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: I guess I was afraid it might have an initial value of null or something, but that's more a concern for notifiers, and it would throw anyway in that case. Just removed the conditional

File: src/util/WalletBackendAdapter.ts (hunk 12 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Added types and left a comment

File: src/util/WalletBackendAdapter.ts (hunk 5 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Looks like `@agoric/patterns` doesn't exist yet either, but that would be cool.

File: src/util/WalletBackendAdapter.ts (hunk 14 above)
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Done

File: src/util/WalletBackendAdapter.ts (hunk 13 above)
Comment (dckc) [reply src/util/WalletBackendAdapter.ts:129]: I was going to say something about the `unknown` type, but if it typechecks (without casts), that means the code doesn't care what the type is, right? i.e. the code is assuming nothing.

File: src/util/queryBankBalances.ts (hunk 15 above)
Comment (samsiegart) [reply src/util/queryBankBalances.ts]: Not sure about the ticket yet, changed to `UNTIL`

File: src/util/queryBankBalances.ts (hunk 3 above)
Comment (dckc) [reply src/util/queryBankBalances.ts:13]: agreed, not a blocker.

meanwhile, I realize pointing to the wiki doesn't automatically make a link the other way, where citing an issue does: https://github.com/Agoric/agoric-sdk/issues/2160

File: src/util/WalletBackendAdapter.ts (hunk 5 above)
Comment (turadg) [reply src/util/WalletBackendAdapter.ts]: Oops I meant '@agoric/store'. But yeah let's wait for it to be in Endo

**Review by turadg — APPROVED** [review]
🚀

File: src/util/queryBankBalances.ts (hunk 4 above)
Comment (dckc) [reply src/util/queryBankBalances.ts:18]: filed:
 - https://github.com/Agoric/agoric-sdk/issues/6870
